File: 238.Product-of-Array-Except-Self.py
# https://leetcode.com/problems/product-of-array-except-self/
#
# algorithms
# Medium (53.88%)
# Total Accepted:    226,840
# Total Submissions: 421,016
# beats 100.0% of python submissions


class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        mul_res = 1
        length = len(nums)
        zero_idx = -1

        for idx, n in enumerate(nums):
            if n == 0:
                if zero_idx == -1:
                    zero_idx = idx
                else:
                    return [0] * length
            else:
                mul_res *= n

        if zero_idx != -1:
            res = [0] * length
            res[zero_idx] = mul_res
            return res

        for idx, n in enumerate(nums):
            nums[idx] = mul_res / n

        return nums

File: 238.Product-of-Array-Except-Self.java
// https://leetcode.com/problems/product-of-array-except-self/
//
// algorithms
// Medium (54.70%)
// Total Accepted:    248,998
// Total Submissions: 455,179
// beats 100.0% of java submissions


class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length, s = 1;
        int res[] = new int[length];
        
        for (int i = 1; i < length; i++) {
            s *= nums[i - 1];
            res[i] = s;
        }
        
        s = 1;
        res[0] = 1;
        
        for (int i = length - 2; i >= 0; i--) {
            s *= nums[i + 1];
            res[i] *= s;
        }
        
        return res;
    }
}